src/transformers/optimization.py

```diff
@@ -165,6 +165,52 @@ def lr_lambda(current_step):
     return LambdaLR(optimizer, lr_lambda, last_epoch)
 
 
+def get_polynomial_decay_schedule_with_warmup(
+    optimizer, num_warmup_steps, num_training_steps, lr_end=1e-5, power=2.0, last_epoch=-1
+):
+    """
+    Create a schedule with a learning rate that decreases as a polynomial decay
+    from the initial lr set in the optimizer to end lr defined by `lr_end`,
+    after a warmup period during which it increases linearly from 0 to the
+    initial lr set in the optimizer.
+
+    Args:
+        optimizer (:class:`~torch.optim.Optimizer`):
+            The optimizer for which to schedule the learning rate.
+        num_warmup_steps (:obj:`int`):
+            The number of steps for the warmup phase.
+        num_training_steps (:obj:`int`):
+            The total number of training steps.
+        lr_end (:obj:`float`, `optional`, defaults to 1e-08):
+            The end LR.
+        power (:obj:`float`, `optional`, defaults to 1.0):
+            Power factor.
+        last_epoch (:obj:`int`, `optional`, defaults to -1):
+            The index of the last epoch when resuming training.
+
+    Return:
+        :obj:`torch.optim.lr_scheduler.LambdaLR` with the appropriate schedule.
+
+    """
+
+    lr_init = optimizer.defaults["lr"]
```

Comment:
assert lr_init > lr_end?
I think we are starting at 3e-5 so 1e-5 feels a little high, but we should just copy fairseq.

Reply:
theirs is 0.0:
``` 
parser.add_argument('--end-learning-rate', default=0.0, type=float)
```

Reply:
I'm also not sure about `lr_init = optimizer.defaults["lr"]` - what if there are groups of `lrs` that aren't the same. The problem with `LambdaLR` - the `base_lr` of the layer isn't passed to `LambdaLR` and therefore this class is not set to deal with different `lrs`.

Reply:
Assert added.